src/video.rs

```diff
@@ -105,68 +118,65 @@ pub fn compare_videos(
-                if let (Some(src_frame), Some(dst_frame)) = (src_frame, dst_frame) {
-                    let src = Yuv::new(src_frame, src_config).unwrap();
-                    let dst = Yuv::new(dst_frame, dst_config).unwrap();
-                    (Rgb::try_from(src).unwrap(), Rgb::try_from(dst).unwrap())
-                } else {
-                    break;
-                }
-            }
-            (_, 8) => {
-                let src_frame = source.read_video_frame::<u16>();
-                let dst_frame = distorted.read_video_frame::<u8>();
-                if let (Some(src_frame), Some(dst_frame)) = (src_frame, dst_frame) {
-                    let src = Yuv::new(src_frame, src_config).unwrap();
-                    let dst = Yuv::new(dst_frame, dst_config).unwrap();
-                    (Rgb::try_from(src).unwrap(), Rgb::try_from(dst).unwrap())
-                } else {
-                    break;
-                }
-            }
-            (_, _) => {
-                let src_frame = source.read_video_frame::<u16>();
-                let dst_frame = distorted.read_video_frame::<u16>();
-                if let (Some(src_frame), Some(dst_frame)) = (src_frame, dst_frame) {
-                    let src = Yuv::new(src_frame, src_config).unwrap();
-                    let dst = Yuv::new(dst_frame, dst_config).unwrap();
-                    (Rgb::try_from(src).unwrap(), Rgb::try_from(dst).unwrap())
+    let (result_tx, result_rx) = mpsc::channel();
+    let src_bd = source.get_bit_depth();
+    let dst_bd = distorted.get_bit_depth();
+
+    let decoders = Arc::new(Mutex::new((source, distorted)));
+    for _ in 0..threads {
+        let decoders = Arc::clone(&decoders);
+        let result_tx = result_tx.clone();
+
+        std::thread::spawn(move || {
+            loop {
+                let score = match (src_bd, dst_bd) {
+                    (8, 8) => calc_score::<u8, u8>(&decoders, &src_config, &dst_config),
+                    (8, _) => calc_score::<u8, u16>(&decoders, &src_config, &dst_config),
+                    (_, 8) => calc_score::<u16, u8>(&decoders, &src_config, &dst_config),
+                    (_, _) => calc_score::<u16, u16>(&decoders, &src_config, &dst_config),
+                };
+
+                if let Some(result) = score {
+                    result_tx.send(result).unwrap();
                 } else {
+                    // no score = no more frames to read
                     break;
                 }
             }
-        };
-        let src_xyb = Xyb::try_from(src_rgb).unwrap();
-        let dst_xyb = Xyb::try_from(dst_rgb).unwrap();
-        let result =
-            compute_frame_ssimulacra2(src_xyb, dst_xyb).expect("Failed to calculate ssimulacra2");
-        if verbose {
-            println!("Frame {frame}: {result:.8}");
-        }
-        results.push(result);
+        });
+    }
+
+    // Needs to be dropped or the main thread never stops waiting for scores
+    drop(result_tx);
```

Comment:
This is safe as far as I'm aware. It looks like it works because `drop` is just an empty function and the variable gets moved into it.

The other way to do this is to wrap your block of code where you want the variable to be alive into its own scope by using curly brackets. It's debatable which one is cleaner, I think the scoping technique is more common possibly.

Reply:
Mhm. I think doing it with a new scope is more common, too. But I like the explicit `drop` more in this case.